### PR TITLE
Capability System: Implicit capability upgrade warning/error

### PR DESCRIPTION
**Pull request by ArielG-NV**
fixes: #4164
fixes: #4165
fixes: #4317

Support a warning/error if capabilities are implicitly upgraded and test accordingly. 

Notable changes:
1. Implicit upgrades are only checked if **Profile/Capability** compiler options are used. Profile is automatically added with render-test in some scenarios.
2. `ImpliesReturnFlags` & `ImpliesFlags` was added to remove `bool` parameters for managing `implies` and to allow for a clearer "implies" operation usage pattern.
3.  `maybeDiagnose` replaces `diagnoseCapabilityErrors` to allow for a more general "type" of error. This was done since "maybeDiagnoseWarningOrError" was added (for togglable warning<->error), and requires a diagnosis wrapper which may not diagnose an error for use with capabilities.
4. `_spirv_*` have been added to handle 'spirv version' capabilities, `spirv_*` are now being used as 'spirv core' (extensions promoted to core are associated with `spirv_*`).
5. `_spirv_*` now translates into `glsl_spirv_*`. This was done since there is not much of a reason to manually specify `glsl_spirv_*` when `spirv` target has a subset of the capabilities of `glsl_spirv_*` (subset since glslang bugs sometimes require inflated versions). 
6. `GLSL_*` and `sm_*` specify a version and cross-platform representation of extensions of the version.

### CODE REVIEW

**Comment by ArielG-NV** [comment]
note: profile is automatically set in render test, likely need to fix capability representation in .capdef (#4163) to avoid errors.

**Review by csyonghe — COMMENTED** [review]
In all of our capability related diagnostic messages, we should try to simplify the atoms so that if `a` implies `b`, we don't print out `b`. For example, if we want to report `spirv_1_5`, we just report `spirv_1_5` instead of `spirv_1_0+spirv_1_1+spirv_1_2+spirv_1_3+spirv_1_4+spirv_1_5`.

**Comment by ArielG-NV** [comment]
> In all of our capability related diagnostic messages, we should try to simplify the atoms so that if a implies b, we don't print out b. For example, if we want to report spirv_1_5, we just report spirv_1_5 instead of spirv_1_0+spirv_1_1+spirv_1_2+spirv_1_3+spirv_1_4+spirv_1_5.

new issue or add to this issue?

File: source/slang/slang-capability.h [hunk 1]
```diff
@@ -139,8 +139,15 @@ struct CapabilitySet
     /// Is this capability set incompatible with the given `other` atomic capability.
     bool isIncompatibleWith(CapabilitySet const& other) const;
 
+    enum class ImpliesReturnFlags : int
+    {
+        NotImplied = 0,
+        Implied = 1 << 1,
```
Comment (kaizhangNV) [inline source/slang/slang-capability.h:145]: 1<<0 or 1<<1 ?

**Comment by csyonghe** [comment]
@ArielG-NV Can do a new issue to simplify capability printing.

File: source/slang/slang-capabilities.capdef [hunk 2]
```diff
@@ -383,120 +405,187 @@ alias fragmentshaderinterlock = _GL_ARB_fragment_shader_interlock | hlsl_nvapi |
...
+                ;
+alias spirv_1_5 = _spirv_1_5 + GL_EXT_nonuniform_qualifier + GL_EXT_buffer_reference
+                | spirv_1_4
+                ;
+alias spirv_1_6 = _spirv_1_6 + GL_EXT_debug_printf + GL_EXT_demote_to_helper_invocation
+                | spirv_1_5
+                ;
+alias spirv_latest = _spirv_1_6;
+
+alias sm_4_0_version = _sm_4_0
```
Comment (csyonghe) [inline source/slang/slang-capabilities.capdef:439]: Maybe add a comment here on what `x_version` means, it means `x` for the corresponding target, and the closest match of equivvalent feature set for other targets.

Comment (csyonghe) [reply source/slang/slang-capabilities.capdef:439]: do we want to automatically translate spirv_1_0 to spirv_1_0_version when the user specifies `slangc shader.slang -profile spirv_1_0_version -target hlsl`?

File: source/slang/slang-capability.cpp [hunk 3]
```diff
@@ -827,6 +848,53 @@ bool CapabilitySet::checkCapabilityRequirement(CapabilitySet const& available, C
...
+inline CapabilityName maybeConvertSpirvVersionToGlslSpirvVersion(CapabilityName& atom)
+{
+    if (atom >= CapabilityName::_spirv_1_0 && asAtom(atom) <= getLatestSpirvAtom())
+    {
+        return (CapabilityName)((Int)CapabilityName::glsl_spirv_1_0 + ((Int)atom - (Int)CapabilityName::_spirv_1_0));
+    }
+    return CapabilityName::Invalid;
+}
+
+void CapabilitySet::AddSpirvVersionFromOtherAsGlslSpirvVersion(CapabilitySet& other)
```
Comment (csyonghe) [inline source/slang/slang-capability.cpp:861]: naming conversion. `Add`->`add`.

File: source/slang/slang-capabilities.capdef (hunk 2 above)
Comment (ArielG-NV) [reply source/slang/slang-capabilities.capdef:439]: > do we want to automatically translate spirv_1_0 to spirv_1_0_version when the user specifies slangc shader.slang -profile spirv_1_0_version -target hlsl?

1. I did not add spirv_*_version but that is a good idea. 

2. We probably also should also warn if a user uses any capability defined as internal in `slangc` (`_*`)

Should '1'/'2' be another issue or part of this PR?

**Comment by csyonghe** [comment]
It seems this is breaking the `slang-unit-test-tool/getTargetCode` test I added yesterday.

**Comment by ArielG-NV** [comment]
> It seems this is breaking the slang-unit-test-tool/getTargetCode test I added yesterday.

I will check it out

**Comment by ArielG-NV** [comment]
`slang-unit-test-tool/getTargetCode` was requesting `-profile sm_5_0` for a spirv target, this causes a warning since sm_5_0 implies spirv_1_0 is requested

**Review by csyonghe — APPROVED** [review]

**Comment by natevm** [comment]
+1 on the simplified compatibility printing. The compatibility system is a good idea, but with too complex of an error printed, I’m more likely to disable the feature than to address it. Programmers are very susceptible to alarm fatigue.

**Comment by ArielG-NV** [comment]
#4347 for tracking the simplified compatibility printing issue

**Pull request merged**